src/rez/version/_requirement.py

```diff
@@ -335,23 +343,24 @@ def _r(r_):
                 r.range_ = range_
                 return r
 
-    def __eq__(self, other):
+    def __eq__(self, other: object) -> bool:
         return (isinstance(other, Requirement)
                 and (self.name_ == other.name_)
                 and (self.range_ == other.range_)
                 and (self.conflict_ == other.conflict_))
 
-    def __hash__(self):
+    def __hash__(self) -> int:
         return hash(str(self))
 
-    def __str__(self):
+    def __str__(self) -> str:
         if self._str is None:
             pre_str = '~' if self.negate_ else ('!' if self.conflict_ else '')
             range_str = ''
             sep_str = ''
 
             range_ = self.range_
-            if self.negate_:
+            # Note: the only time that range_ is None is if self.negate_ is True
+            if self.negate_ or range_ is None:
```

Comment:
I'm not sure this is the right thing to do. Why should `range_` be None?

Reply:
this is the init code:

```python
        m = self.sep_regex.search(s)
        if m:
            i = m.start()
            self.name_ = s[:i]
            req_str = s[i:]
            if req_str[0] in ('-', '@', '#'):
                self.sep_ = req_str[0]
                req_str = req_str[1:]
            self.range_ = VersionRange(
                req_str, invalid_bound_error=invalid_bound_error)
            if self.negate_:
                self.range_ = ~self.range_
        elif self.negate_:
            self.name_ = s
            # rare case - '~foo' equates to no effect
            self.range_ = None
        else:
            self.name_ = s
            self.range_ = VersionRange()
```

Notice that if `self.negate_ `is True, `self._range_` is not set.